AsyncDisplayKit/ASViewController.mm

```diff
@@ -132,4 +142,70 @@ - (ASInterfaceState)interfaceState
   return _node.interfaceState;
 }
 
+#pragma mark - ASEnvironmentTraitCollection
+
+- (ASEnvironmentTraitCollection)displayTraitsForTraitCollection:(UITraitCollection *)traitCollection
+{
+  if (self.overrideDisplayTraitsWithTraitCollection) {
+    ASTraitCollection *asyncTraitCollection = self.overrideDisplayTraitsWithTraitCollection(traitCollection);
+    asyncTraitCollection.isMutable = NO;
+    return [asyncTraitCollection environmentTraitCollection];
+  }
+  
+  ASEnvironmentTraitCollection asyncTraitCollection = ASEnvironmentTraitCollectionFromUITraitCollection(traitCollection);
+  asyncTraitCollection.displayContext = _displayTraitsContext;
+  return asyncTraitCollection;
+}
+
+- (ASEnvironmentTraitCollection)displayTraitsForWindowSize:(CGSize)windowSize
+{
+  if (self.overrideDisplayTraitsWithWindowSize) {
+    ASTraitCollection *traitCollection = self.overrideDisplayTraitsWithWindowSize(windowSize);
+    traitCollection.isMutable = NO;
+    return [traitCollection environmentTraitCollection];
+  }
+  return self.node.environmentState.traitCollection;
+}
+
+- (void)progagateNewDisplayTraits:(ASEnvironmentTraitCollection)traitCollection
```

Comment:
Oh, you are talking about moving `progagateNewDisplayTraits:` to the node level. That is possible. The VC will still be the one to initiate the propagation since it is the one that knows when the trait collection actually changed. Personally, I like that all the trait collection details are encapsulated inside of the VC, but I definitely see your point.